ui/components/DagGraph.tsx

```diff
@@ -0,0 +1,205 @@
+import _ from "lodash";
+import * as React from "react";
+import * as d3d from "d3-dag";
+import styled from "styled-components";
+import { Slider } from "@material-ui/core";
+import { FluxObjectNode } from "../lib/objects";
+import Flex from "./Flex";
+import DagGraphNode from "./DagGraphNode";
+import Spacer from "./Spacer";
+
+type NodeSize = {
+  width: number;
+  height: number;
+  verticalSeparation: number;
+  horizontalSeparation: number;
+  arrowWidth: number;
+  arrowHeight: number;
+};
+
+type Props = {
+  className?: string;
+  nodes: FluxObjectNode[];
+};
+
+const SliderFlex = styled(Flex)`
+  padding-top: ${(props) => props.theme.spacing.base};
+  min-height: 400px;
+  min-width: 60px;
+  width: 5%;
+`;
+
+const PercentFlex = styled(Flex)`
+  color: ${(props) => props.theme.colors.primary10};
+  padding: 10px;
+  background: rgba(0, 179, 236, 0.1);
+  border-radius: 2px;
+`;
+
+const GraphDiv = styled.div`
+  width: 100%;
+  height: 100%;
+`;
+
+function DagGraph({ className, nodes }: Props) {
+  //zoom
+  const defaultZoomPercent = 85;
+  const [zoomPercent, setZoomPercent] = React.useState(defaultZoomPercent);
+
+  //pan
+  const [pan, setPan] = React.useState({ x: 0, y: 0 });
+  const [isPanning, setIsPanning] = React.useState(false);
+  const handleMouseDown = () => {
+    setIsPanning(true);
+  };
+  const handleMouseMove = (e) => {
+    //viewBox change. e.movement is change since previous mouse event
+    if (isPanning) setPan({ x: pan.x + e.movementX, y: pan.y + e.movementY });
+  };
+  const handleMouseUp = () => {
+    setIsPanning(false);
+  };
+
+  //minimum zoomBox is 1000
+  const zoomBox = 15000 - 14000 * (zoomPercent / 100);
+  //since viewbox is so large, make smaller mouse movements correspond to larger pan
+  const svgRef = React.useRef<SVGSVGElement>();
+  let panScale = 1;
+  if (svgRef.current) {
+    panScale = zoomBox / svgRef.current.getBoundingClientRect().width;
+  }
+
+  //graph numbers
+  const nodeSize: NodeSize = {
+    width: 800,
+    height: 300,
+    verticalSeparation: 150,
+    horizontalSeparation: 100,
+    arrowWidth: 25,
+    arrowHeight: 25,
+  };
+
+  const arrowHalfWidth = nodeSize.arrowWidth / 2;
+  const nodeHalfHeight = nodeSize.height / 2;
+  const linkStrokeWidth = 5;
+
+  //use d3 to create DAG structure
```

Comment:
Maybe we could move the shared functionality like zooming/dragging/panning to a hook.

Reply:
And the API calls to d3 or d3-dag libraries are different in both graphs.

Reply:
So, there will be constant branching logic "if this graph type then...".

Usually, AFAIK< different chart types are not combines into a single component.

Reply:
The difference seems to be:
* Shape of arrows (needs conditionals?)
* Component to render objects (can be provided)
* Right now, the the input to `ReconciliationGraph` takes a tree for d3-hierarchy while the input to `DagGraph` takes a flat list for d3-dag.

I think I would drop `d3-hierarchy` entirely, and change the existing graph to generate a flat list of objects again, so that `DagGraph` can render all graphs. 
 
However. I think it's a good idea to do that Another Day™ - this is a lot of new code, so let's wait a couple of releases before we decide if the old code should be migrated to this or not.

Reply:
OK, sounds good to me. I could try to combine the graphs now but maybe Product decides that they want the new one to look different. And if it requires serious rework of the existing graph component, I think that yes, it's better to do it in a separate ticket.

Another difference between the graphs is offset to shift the graph horizontally to center it. And I somewhat changed the way the lines are drawn int he new graph (int he old graph they are draw between node centers and pass under the nodes what is not needed in general, but it can be combined). The main diff is input data format, yes.

I guess, ReconciliationGraph should be a think wrapper component. All non-reconciliation-specific functionality should be moved to the DirectedGraph anyway even if we don't combine the graphs.

Reply:
Arrows were not drawn in the original graph, they were added as a separate path in the new one.